### PR TITLE
CBG-1884: Documented attachment endpoints

### PR DESCRIPTION
**Pull request by IsaacLambat**
CBG-1884

Endpoints documented:
`GET HEAD PUT /{db}/{docid}/{attach}`

## Dependencies
- [x] Merge after #5410

### CODE REVIEW

File: docs/api/api_docs.yaml [hunk 1]
```diff
@@ -976,27 +976,111 @@ paths:
...
+        '416':
+          description: Requested range exceeds content length
       tags:
         - Admin
         - Public
+      summary: Get an attachment from a document
+      description: |-
+        This request retrieves a file attachment associated with the document. 
+
+        The raw data of the associated attachment is returned (just as if you were accessing a static file). The `Content-Type` response header is the same content type set when the document attachment was added to the database. The `Content-Disposition` response header will be set instead if the content type is vulnerable to a phishing attack, causing the browser to download the file instead of display it (unless overriden by by database config option `serve_insecure_attachment_types`).
```
Comment (bbrks) [inline docs/api/api_docs.yaml]:
```suggestion
        The raw data of the associated attachment is returned (just as if you were accessing a static file). The `Content-Type` response header is the same content type set when the document attachment was added to the database. The `Content-Disposition` response header to force attachment downloads will be set if the content type is considered unsafe to display in a browser (unless overriden by by database config option `serve_insecure_attachment_types`).
```

File: docs/api/api_docs.yaml [hunk 2]
```diff
@@ -976,27 +976,111 @@ paths:
...
+
+        If the `meta` query parameter is set then the response will be in JSON with the additional metadata tags.
+      parameters:
+        - $ref: '#/components/parameters/rev'
+        - schema:
+            type: boolean
+            default: 'false'
+          in: query
+          name: meta
+          description: Include the metadata of the attachment in the response body. Will cause `Content-Type=application/json` in the response header.
```
Comment (bbrks) [inline docs/api/api_docs.yaml]:
```suggestion
          description: Return only the metadata of the attachment in the response body.
```

File: docs/api/api_docs.yaml [hunk 3]
```diff
@@ -976,27 +976,111 @@ paths:
...
         - Public
+      summary: Create or update an attachment on a document
+      description: |-
+        This request adds or updates an attachment associated with the document. If the document does not exist, it will be created and the attachment will be added to it.
+
+        The maximum content size of an attachment is 20MB. The `Content-Type` header of the request should be what the content type of the attachment is otherwise this will default to `application/octet-stream`.
+
+        If the attachment already exists, the data of the existing attachment will be replaced in the new revision. The revisions are for tracking document changes, not attachment changes.
+
+        To remove an attachment from a document, simply update the `_attachments` dictionary of the document in the `PUT /{db}/{docid}` request.
```
Comment (bbrks) [inline docs/api/api_docs.yaml]:
```suggestion
        To remove an attachment from a document, omit the attachment in the `_attachments` object of the document in the `PUT /{db}/{docid}` request.
```

File: docs/api/api_docs.yaml [hunk 4]
```diff
@@ -976,27 +976,111 @@ paths:
...
       tags:
         - Admin
         - Public
+      summary: Create or update an attachment on a document
+      description: |-
+        This request adds or updates an attachment associated with the document. If the document does not exist, it will be created and the attachment will be added to it.
+
+        The maximum content size of an attachment is 20MB. The `Content-Type` header of the request should be what the content type of the attachment is otherwise this will default to `application/octet-stream`.
+
+        If the attachment already exists, the data of the existing attachment will be replaced in the new revision. The revisions are for tracking document changes, not attachment changes.
```
Comment (bbrks) [inline docs/api/api_docs.yaml]: Swap these two paragraps around to keep the high-level description at the top. Specifics around headers/size limits further down.

Comment (bbrks) [inline docs/api/api_docs.yaml]: > The revisions are for tracking document changes, not attachment changes.

I'm not really sure what this means. Attachment changes are included as part of a document change.

File: docs/api/api_docs.yaml [hunk 5]
```diff
@@ -976,27 +976,111 @@ paths:
...
+        '409':
+          $ref: '#/components/responses/Conflict'
       tags:
         - Admin
         - Public
+      summary: Create or update an attachment on a document
+      description: |-
+        This request adds or updates an attachment associated with the document. If the document does not exist, it will be created and the attachment will be added to it.
+
+        The maximum content size of an attachment is 20MB. The `Content-Type` header of the request should be what the content type of the attachment is otherwise this will default to `application/octet-stream`.
```
Comment (bbrks) [inline docs/api/api_docs.yaml]:
```suggestion
        The maximum content size of an attachment is 20MB. The `Content-Type` header of the request specifies the content type of the attachment.
```

Reword and default values are described already in the spec. I don't think there's a need to repeat them in a description.

File: docs/api/api_docs.yaml [hunk 6]
```diff
@@ -976,27 +976,111 @@ paths:
...
+        - name: rev
+          schema:
+            type: string
+          in: query
+          description: The document revision to add/modify the attachment on. Required only when modifying an existing document. This falls back on the `If-Match` header if not set.
+        - name: If-Match
+          schema:
+            type: string
+          in: header
+          description: Used as the `rev` query parameter if it is not set.
```
Comment (bbrks) [inline docs/api/api_docs.yaml]: Wording of this is ambiguous and exposes some unnecessary implementation detail.

```suggestion
          description: The existing document revision ID to modify. Required only when modifying an existing document.
        - name: If-Match
          schema:
            type: string
          in: header
          description: An alternative way of specifying the document revision ID.
```

File: docs/api/api_docs.yaml [hunk 7]
```diff
@@ -976,27 +976,111 @@ paths:
           type: string
         in: path
         required: true
+        description: 'The attachment name. This value must be URL encoded. For example, if the attachment name is `blob_/avatar`, the path component passed to the URL should be `blob_%2Favatar` (tested with [URLEncoder](https://www.urlencoder.org/)).'
     get:
       responses:
         '200':
-          description: OK
+          description: Found attachment successfully.
+          headers:
```
Comment (bbrks) [inline docs/api/api_docs.yaml:984]:
```suggestion
          headers:
            Content-Length:
              schema:
                type: string
                description: length of attachment in bytes
```

File: docs/api/api_docs.yaml [hunk 8]
```diff
@@ -976,27 +976,111 @@ paths:
...
         - Public
+      summary: Get an attachment from a document
+      description: |-
+        This request retrieves a file attachment associated with the document. 
+
+        The raw data of the associated attachment is returned (just as if you were accessing a static file). The `Content-Type` response header is the same content type set when the document attachment was added to the database. The `Content-Disposition` response header will be set instead if the content type is vulnerable to a phishing attack, causing the browser to download the file instead of display it (unless overriden by by database config option `serve_insecure_attachment_types`).
+
+        If the `meta` query parameter is set then the response will be in JSON with the additional metadata tags.
+      parameters:
+        - $ref: '#/components/parameters/rev'
```
Comment (bbrks) [inline docs/api/api_docs.yaml:1010]: Missing request parameters

```suggestion
        - $ref: '#/components/parameters/rev'
        - schema:
            type: boolean
            default: 'true'
          in: query
          name: content_encoding
          description: Set to false to disable the `Content-Encoding` response header.
        - schema:
            type: string
          in: header
          name: Range
          description: 'RFC-2616 bytes range header. E.g: bytes=123-456'
```

File: docs/api/api_docs.yaml [hunk 9]
```diff
@@ -976,27 +976,111 @@ paths:
...
+            schema:
+              description: The content to store in the body
+              type: string
+        description: The attachment data
     head:
       responses:
         '200':
-          description: OK
+          description: The document exists and the attachment exists on the document.
+          headers:
```
Comment (bbrks) [inline docs/api/api_docs.yaml:1086]: Content-Length

**Review by bbrks — CHANGES_REQUESTED** [review]

File: docs/api/api_docs.yaml (hunk 4 above)
Comment (IsaacLambat) [reply docs/api/api_docs.yaml]: I was trying to say that attachments don't get their own revision ID but rather the document does. Will remove this as it seems unnecessary.

File: docs/api/api_docs.yaml [hunk 10]
```diff
@@ -982,6 +982,10 @@ paths:
         '200':
           description: Found attachment successfully.
           headers:
+            Content-Length:
+              schema:
+                type: number
+                description: Length of the attachment in bytes
```
Comment (bbrks) [inline docs/api/api_docs.yaml]: whoops, my suggestion was incorrectly indented

```suggestion
              description: Length of the attachment in bytes
```

File: docs/api/api_docs.yaml [hunk 11]
```diff
@@ -999,17 +1003,28 @@ paths:
...
+          schema:
+            type: boolean
+            default: 'true'
+          in: query
+          description: 'Set to false to disable the `Content-Encoding` response header.'
+        - name: Range
+          schema:
+            type: string
+          in: header
+          description: 'RFC-2616 bytes range header. For example, bytes=123-456.'
```
Comment (bbrks) [inline docs/api/api_docs.yaml]: Again my bad w/ suggestion - I've seen the correct way of specifying an example:

```suggestion
          description: 'RFC-2616 bytes range header.'
          example: 'bytes=123-456'
```

**Review by bbrks — CHANGES_REQUESTED** [review]
Feel free to just apply these suggestions in GitHub - they were my incorrect suggestions to begin with 😄

**Review by bbrks — APPROVED** [review]

**Pull request merged**